### PR TITLE
Game dev hoc

### PR DESCRIPTION
**Pull request by sderickson**
Updates to the game for the Game Dev Hour of Code activity.

* Make accessible all game-dev levels and game-dev db functionality to anonymous HoC players
* Further suppress signup and payment prompts
* UI bug fixes and updates

### CODE REVIEW

File: app/views/play/level/modal/HeroVictoryModal.coffee [hunk 1]
```diff
@@ -416,7 +416,8 @@ module.exports = class HeroVictoryModal extends ModalView
     campaign = @level.get 'campaign'
     if @level.get('slug') in campaignEndLevels
       campaign = ''  # Return to campaign selector
-    if (campaign is 'dungeon' or @level.get('slug') in ['kithgard-gates', 'game-grove']) and storage.load('should-return-to-game-dev-hoc')
+    gdHocLevels = ['kithgard-gates', 'over-the-garden-wall', 'vorpal-mouse', 'forest-incursion', 'them-bones', 'behavior-driven-development', 'seeing-is-believing', 'persistence-pays', 'game-grove']
```
Comment (differentmatt) [inline app/views/play/level/modal/HeroVictoryModal.coffee]: Are these level names locked down in practice?  Should use ids if not.

File: app/views/play/level/modal/HeroVictoryModal.coffee [hunk 2]
```diff
@@ -416,7 +416,8 @@ module.exports = class HeroVictoryModal extends ModalView
...
+    if (campaign is 'dungeon' or @level.get('slug') in gdHocLevels) and storage.load('should-return-to-game-dev-hoc')
```
Comment (differentmatt) [inline app/views/play/level/modal/HeroVictoryModal.coffee]: Should this use the new `application.onHocPage` function?

File: app/views/play/CampaignView.coffee [hunk 3]
```diff
@@ -479,14 +479,15 @@ module.exports = class CampaignView extends RootView
...
-    return if features.noAuth
+    return if features.noAuth or @terrain is 'game-dev-hoc'
 
     @endHighlight()
     authModal = new CreateAccountModal supermodel: @supermodel
     authModal.mode = 'signup'
     @openModalView authModal
 
   promptForSubscription: (slug, label) ->
+    return console.log('Game dev HoC does not encourage subscribing.') if @terrain is 'game-dev-hoc'
```
Comment (differentmatt) [inline app/views/play/CampaignView.coffee]: Who is this console.log for?

File: app/views/play/CampaignView.coffee [hunk 4]
```diff
@@ -754,7 +756,11 @@ module.exports = class CampaignView extends RootView
     levelURL = "/db/level/#{levelSlug}"
     level = new Level().setURL levelURL
     level = @supermodel.loadModel(level, null, 0).model
-    sessionURL = "/db/level/#{levelSlug}/session"
+
+    # Note that this not only preloads the level, but also creates it if the
+    # level requires the campaign to be included, which the game-dev-hoc premium levels.
```
Comment (differentmatt) [inline app/views/play/CampaignView.coffee]: `Note this preloads the level, creates ?? if the level requires the campaign to be included.  E.g. game-dev-hoc premium levels.`

File: server/middleware/levels.coffee [hunk 5]
```diff
@@ -124,6 +125,18 @@ module.exports =
...
         req.features.codePlay and codePlay.canPlay(level.get('slug'))
       ])
+
+      if req.query.campaign and not canPlayAnyway
+        # check if the campaign requesting this is game dev hoc, if so then let it work
+        query = {
+          _id: mongoose.Types.ObjectId(req.query.campaign),
+          slug: 'game-dev-hoc', # replace with a way of indicating that 
+        }
+        query['levels.'+level.get('original')] = {$exists: true}
```
Comment (differentmatt) [inline server/middleware/levels.coffee]: Why is this a separate line?

File: app/views/play/level/modal/HeroVictoryModal.coffee (hunk 1 above)
Comment (sderickson) [reply app/views/play/level/modal/HeroVictoryModal.coffee]: Yes, they're as locked down as the rest of the levels we do this for. It's expected, I believe, that if we change names of published levels that the codebase needs to be checked for hardcodings like this. Would be good to figure out a better system, but that's a larger project.

File: app/views/play/level/modal/HeroVictoryModal.coffee (hunk 2 above)
Comment (sderickson) [reply app/views/play/level/modal/HeroVictoryModal.coffee]: Good point!

File: server/middleware/levels.coffee (hunk 5 above)
Comment (sderickson) [reply server/middleware/levels.coffee]: Cause dynamic keys + object literals don't mix, or so I thought. Can't in JS, but looks like possible in CS. Tweaked it.

File: app/views/play/CampaignView.coffee (hunk 3 above)
Comment (sderickson) [reply app/views/play/CampaignView.coffee]: Anyone who might try to be figuring out why CampaignView is acting strangely.

File: app/views/play/CampaignView.coffee (hunk 4 above)
Comment (sderickson) [reply app/views/play/CampaignView.coffee]: Clarified comment:

```
    # Note that this doesn't just preload the level. For sessions which require the
    # campaign to be included, it also creates the session. If this code is changed,
    # make sure to accommodate campaigns with free-in-certain-campaign-contexts levels,
    # such as game dev levels in game-dev-hoc.
```

**Pull request merged**